Extend IOContext to reader/writer/seeker

BubbaJoe opened the pull request:
## Purpose
Allows you to read/write/seek data that you already have in memory
Allows you to read from a file a file via IOContext

asticode commented on io_context.go:
```diff
@@ -3,11 +3,16 @@ package astiav
 // https://github.com/FFmpeg/FFmpeg/blob/n5.0/libavformat/avio.h#L161
 type IOContext struct {
 	c *C.struct_AVIOContext
+	io.ReadWriteCloser
```
Can you remove those 2 lines?

BubbaJoe replied:
Actually i wanted to use IOContext as an io.Writer / io.Reader.

So i can pass data in memory directly to it.

asticode replied:
Sure, but there's no need to declare interfaces here, the simple fact that IOContext would implement those interfaces is sufficient

asticode commented on io_context.go:
```diff
@@ -25,15 +30,36 @@ func (ic *IOContext) Closep() error {
+func (ic *IOContext) Read(b []byte) (n int, err error) {
+	n = int(C.avio_read(ic.c, (*C.uchar)(unsafe.Pointer(&b[0])), C.int(len(b))))
+	if n < 0 {
+		return 0, io.EOF
```
I'd rather you return `0, newError(n)` instead

BubbaJoe replied:
i think `io.Reader` uses io.EOF so that is knows when to stop reading. Let me try it out though.

asticode replied:
Thing is, I want `astiav` to be as close as possible to C since it's merely an expression of the C code. Implementing `io.ReadWriteCloser` as well as the `io.Seeker` interfaces seems like a good idea but I'm wondering whether we should create a wrapper on top of `IOContext` instead 🤔 

Things that are bothering me:

- the `Close` method that would implement the `io.Closer` interface should use `Closep()` instead
- the `Write` method shouldn't return `(n int)` since C doesn't provide that information
- the `Read` method should return an `astiav` error

Here's what I'm proposing:

- update `IOContext` methods you've added so that they're closer to C (cf. my other comments)
- create an `IOContextWrapper` struct (let me know if you think of a better name) that would only implement the `io.ReadWriteCloser` as well as the `io.Seeker` interfaces (in those methods, you could return `io.EOF` for instance)
- add an `IOContext`.`Wrapper()` method that would return an `IOContextWrapper`

That people that want to stay close to C can use `IOContext`, and people that want to use GO interfaces can use `IOContextWrapper`

BubbaJoe replied:
Thanks for the proposal. I am currently looking at https://ffmpeg.org/doxygen/5.0/structAVIOContext.html

And it seems that there are some fields that I can use to extract information about how many bytes were written; testing in progress.

Concerning, the `io.Closer`, i actually just realized that we probably shouldn't implement it as `avio_closep(...)` (and `avio_close(...)`) is only needed when the resource is opened by`avio_open(...)`

I can manually check [eof_reached](https://ffmpeg.org/doxygen/5.0/structAVIOContext.html#aa9214f54c01f07076e0a0718b2e9c27b), and return `io.EOF` as the error.

What do you think? Do you still think a wrapper is better? I think the API will be confusing/weird with wrappers.

FYI: [Read needs to implement io.EOF](https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/io/io.go;l=44)

asticode replied:
> I think the API will be confusing/weird with wrappers.

I agree with you.

I don't think implementing the `io.ReadWriteCloser` and `io.Seeker` is a good idea in this lib since its main mission is to be the closest possible to C. If somebody wants to implement this interface, he/she can do it in their project, it's pretty straight-forward.

Therefore here's what I'd rather have in this PR:

- add `func (ic *IOContext) Close() error`
- keep `Write()` as it is
- add `func (ic *IOContext) Read(b []byte) (n int, err error)`
- add `func (ic *IOContext) Seekable() bool`
- add `func (ic *IOContext) Seek(offset int64, whence int) (int64, error)`
- add `IOContext` getters if you need them to implement the `io` interfaces (such as `bytes_written` for instance)

and that's it. That should provide the proper method to anybody looking to implement `io.ReadWriteCloser` and `io.Seeker`.

BubbaJoe replied:
Makes sense will do.

asticode commented on io_context.go:
```diff
@@ -25,15 +30,36 @@ func (ic *IOContext) Closep() error {
+		return 0, nil
 	}
 	C.avio_write(ic.c, (*C.uchar)(unsafe.Pointer(&b[0])), C.int(len(b)))
+	return len(b), nil
```
This is misleading since we don't really know if the amount of written bytes is `len(b)` 🤔

BubbaJoe replied:
Yea, i realized this too while coding it, but didn't think it was a big deal. I think there might be a way to tell but i dont quite understand the C code here.

https://ffmpeg.org/doxygen/5.0/aviobuf_8c_source.html#l00232

asticode commented on io_context.go:
```diff
@@ -25,15 +30,36 @@ func (ic *IOContext) Closep() error {
+}
+
+func (ic *IOContext) Read(b []byte) (n int, err error) {
+	n = int(C.avio_read(ic.c, (*C.uchar)(unsafe.Pointer(&b[0])), C.int(len(b))))
```
Could you not cast this to `int()` right away ?

asticode commented on io_context.go:
```diff
@@ -25,15 +30,36 @@ func (ic *IOContext) Closep() error {
+	if n < 0 {
+		return 0, io.EOF
+	}
+	return n, nil
```
You can cast n to int() here

asticode commented on io_context.go:
```diff
@@ -25,15 +30,36 @@ func (ic *IOContext) Closep() error {
+}
+
+func (ic *IOContext) Seek(offset int64, whence int) (int64, error) {
+	return int64(C.avio_seek(ic.c, C.int64_t(offset), C.int(whence))), nil
```
Can you use a logic similar to this instead : 

```go

ret := C.avio_seek(ic.c, C.int64_t(offset), C.int(whence))
if ret < 0 {
  return 0, newError(C.int(ret))
}
return int64(ret), nil
```

asticode commented on io_context_test.go:
```diff
@@ -1,27 +1,133 @@
 	"github.com/stretchr/testify/require"
 )
 
-func TestIOContext(t *testing.T) {
```
As the IOContext tests are not that big, can you avoid splitting them into separate functions?

For instance you could:

- Write to the file using the IOContext => check error + n
- Seek using the IOContext
- Read from the file using the IOContext => check error + n + string == "testing"

Also, could you test `Seekable` as well?

BubbaJoe replied:
Sure, but concerning `Seekable`; im not sure if what case it would be false. Let me know if you have any suggestions for that.

asticode replied:
I think testing it's `true` when opening a file is sufficient 👍

asticode reviewed (changes requested):
This is a nice idea event though the `Write` method is a bit misleading 🤔

asticode commented:
You definitely need to alloc the context manually if you don't use `Open()`. But that's a pretty hard change to do with callbacks and free() constraints. I wouldn't do it if I were you and stick with Open() for testing

BubbaJoe commented:
Hello, Sorry for the delay; got distracted by all of the new games 😵‍💫,

After your comments i realized that i was implementing everything wrong. So i re-did everything from the start. The callback functionality needed a bit of hacking to get working, but it looks 10 times cleaner than my first iteration of it. Anyways, I think this implementation is idiomatic to work seamlessly with go standards, but not too much because you can still use the original C functions directly.

One issue that I am still having is that once I create the IOContext using `avio_alloc_context(...)`; I cannot use `(*IOContext) Close()` or  `(*IOContext) Closep()` to close, as it results in a segmentation violation. However, `(*IOContext) Free()` is working and i am using that in tests. I am thinking of using `(*IOContext) Free()` when `(*IOContext) Close()` is called in the case that i use `avio_alloc_context`.

Please try out the tests i added:
go test -benchmem -run ^TestIOContext -v 

Please try out the benchmarks i added:
`go test -benchmem -run ^$ -bench '^BenchmarkIOContext' -cpu 1,2,4,8,16`


Additionally, I needed support for sws_scale (#8), so I implemented this but i will put it in another PR as soon I add proper testing.

I have created a cool (gocv example)[https://github.com/BubbaJoe/go-astiav-contr/tree/misc-update/examples/gocv], let me know if you are interesting in having examples that are different from original ffmpeg examples. Thanks!

BubbaJoe commented:
@asticode
